### PR TITLE
Cut over from SearchContext to ReaderContext

### PR DESCRIPTION
**Pull request by dnhatn**
With this change, we partially move the state of SearchContext to ReaderContext. This is another step allowing us to move the state of search to the coordinating node.

We will need several follow-ups to move the entire search state to the coordinating node.

Relates #46523 

Note: this pull request targets the `reader-context` branch.

### CODE REVIEW

**Review by jimczi — APPROVED** [review]
I left one question and an idea for a possible simplification. Other than that the pr looks great and we can try to optimize the parsing in a follow up (lazy rewrite of the query in the fetch phase for instance).

File: server/src/main/java/org/elasticsearch/search/internal/ReaderContext.java [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    private final AtomicBoolean closed = new AtomicBoolean(false);
+
+    private volatile long keepAlive;
+    private volatile long lastAccessTime = -1L;
+
+    // BWC
+    private final ShardSearchRequest request;
+    private final ScrollContext scrollContext;
+    private AggregatedDfs aggregatedDfs;
+    private List<RescoreContext> rescore;
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/search/internal/ReaderContext.java]: We set [rescoredDocIds](https://github.com/elastic/elasticsearch/blob/b9c98c3493ee91f85a405dec5025a6ea06f072b7/server/src/main/java/org/elasticsearch/search/rescore/QueryRescorer.java#L71) in the query phase and use them in the [explain](https://github.com/elastic/elasticsearch/blob/804836dad75c006e8ee3590901e04fc4dbd7f01c/server/src/main/java/org/elasticsearch/search/rescore/QueryRescorer.java#L97) of the fetch phase . If we pass the rescoredDocIds around, then we can remove this.

File: server/src/main/java/org/elasticsearch/search/internal/SearchContext.java [hunk 2]
```diff
@@ -411,10 +377,6 @@ public final boolean hasOnlySuggest() {
          * This life time is for objects that only live during collection time.
          */
         COLLECTION,
-        /**
-         * This life time is for objects that need to live until the end of the current search phase.
-         */
-        PHASE,
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/search/internal/SearchContext.java]: ++. I removed it in https://github.com/elastic/elasticsearch/pull/51282/commits/34ed39bcae42806fb5c016884f1a56ad504c9af6.

**Comment by dnhatn** [comment]
@jimczi Thank you for your review :).

**Pull request merged**